### PR TITLE
Server implementation

### PR DESCRIPTION
**Pull request by FONARMEISTER**

### CODE REVIEW

File: Elysium/.gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+HELP.md
```
Comment (devcrocod) [inline Elysium/.gitignore]: What is HELP.md?

File: Elysium/build.gradle [hunk 2]
```diff
@@ -1,34 +1,42 @@
...
+group = 'hse.elysium'
+version = '0.0.1-SNAPSHOT'
+sourceCompatibility = '17'
 
 repositories {
     mavenCentral()
 }
-
 ext {
     junitVersion = '5.9.1'
```
Comment (devcrocod) [inline Elysium/build.gradle:15]: Versions can be stored in gradle.properties

File: Elysium/build.gradle [hunk 3]
```diff
@@ -1,34 +1,42 @@
...
+test {
+    useJUnitPlatform()
+}
+
 dependencies {
-    implementation('org.hibernate:hibernate-core:6.0.2.Final')
-    implementation('org.glassfish.jaxb:jaxb-runtime:3.0.2')
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.hibernate:hibernate-core:6.0.2.Final'
+    implementation 'org.glassfish.jaxb:jaxb-runtime:3.0.2'
```
Comment (devcrocod) [inline Elysium/build.gradle:30]: What is glassfish.jaxb used for in this project?

File: Elysium/src/main/java/hse/elysium/databaseInteractor/TokenService.java [hunk 4]
```diff
@@ -3,9 +3,11 @@
...
 
 import hse.elysium.entities.Token;
+import org.springframework.stereotype.Service;
 
 import java.util.ArrayList;
 
+@Service
 public class TokenService {
     private final EntityManagerFactory
         entityManagerFactory = Persistence.createEntityManagerFactory("default");
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/databaseInteractor/TokenService.java:13]: Couldn't you achieve it without native SQL queries and with type safety?

File: Elysium/src/main/java/hse/elysium/databaseInteractor/UserService.java [hunk 5]
```diff
@@ -307,6 +312,17 @@ public int deleteTrackFromFavouritesWithUserId(int user_id, int track_id) {
...
     }
 
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        int userId = getUserIdWithLogin(username);
+        if (userId == -1){
+            throw new UsernameNotFoundException("No such login");
+        }
+        User user = getUserWithUserId(userId);
+        return new org.springframework.security.core.userdetails.User(
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/databaseInteractor/UserService.java:322]: import?

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    /*
+        a.k.a. authentication
+     */
+    @PostMapping("/login")
+    public ResponseEntity<String> login(@RequestBody UserLoginPasswordForm userRegisterForm) {
+        return ResponseEntity.ok(authenticationService.login(userRegisterForm.getLogin(), userRegisterForm.getPassword()));
+    }
+
+    @RequestMapping("/bebra")
+    public Integer bebra() {
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java]: ?

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@RequestMapping("/elysium/auth")
+@RequiredArgsConstructor
+@ComponentScan("hse.elysium")
+public class AuthController {
+
+    private final AuthenticationService authenticationService;
+
+    @PostMapping("/register")
+    public ResponseEntity<String> register(@RequestBody UserLoginPasswordForm userRegisterForm) {
+        String res = authenticationService.register(userRegisterForm.getLogin(), userRegisterForm.getPassword());
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java:19]: Why is the response a string?

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+@ComponentScan("hse.elysium")
+public final class AuthenticationFilter extends OncePerRequestFilter {
+
+    private final JwtService jwtService;
+    private final UserDetailsService userDetailsService;
+    private final TokenService tokenService;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        System.out.println("AuthenticationFilter.doFilterInternal");
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/AuthenticationFilter.java]: It's better to use logging, and where is the output directed to?

File: Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import java.security.Key;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+@Service
+public class JwtService {
+
+    private static final String SECRET_KEY = "4A614E645267556B58703273357638792F413F4428472B4B6250655368566D59";
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java]: Secret key?))))

File: Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        return extractClaim(token, Claims::getIssuer);
+    }
+
+    public <T> T extractClaim(String token, Function<Claims, T> claimsResolver) {
+        final Claims claims = extractAllClaims(token);
+        return claimsResolver.apply(claims);
+    }
+
+    public String generateToken(User user) {
+        return generateToken(new HashMap<>(), user);
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java:59]: Why is a new map passed in?
Is `generateToken` called somewhere with an existing map?

File: Elysium/src/main/java/hse/elysium/serverspring/auth/UserLoginPasswordForm.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class UserLoginPasswordForm {
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/auth/UserLoginPasswordForm.java:12]: What is purpose of this class?

File: Elysium/src/main/java/hse/elysium/serverspring/config/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    private static final RequestMatcher PROTECTED_URLS = new NegatedRequestMatcher(PUBLIC_URLS);
+
+    private final AuthenticationFilter authFilter;
+    private final AuthenticationProvider authenticationProvider;
+    private final LogoutHandler logoutHandler;
+
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
```
Comment (devcrocod) [inline Elysium/src/main/java/hse/elysium/serverspring/config/SecurityConfig.java:43]: What exceptions are thrown here exactly? And it's better to fix indentation

**Review by devcrocod — APPROVED** [review]

File: Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java (hunk 9 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java]: fixed

File: Elysium/src/main/java/hse/elysium/serverspring/auth/UserLoginPasswordForm.java (hunk 11 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/UserLoginPasswordForm.java:12]: To describe json for register/login

File: Elysium/.gitignore (hunk 1 above)
Comment (FONARMEISTER) [reply Elysium/.gitignore]: There are no such file in directory, but it surely shouldn't be in gitignore

File: Elysium/build.gradle (hunk 2 above)
Comment (FONARMEISTER) [reply Elysium/build.gradle:15]: internet says ext is also a good place to store

File: Elysium/build.gradle (hunk 3 above)
Comment (FONARMEISTER) [reply Elysium/build.gradle:30]: It was autogenerated, I don't know for sure, but Internet says it something to work with XML.

File: Elysium/src/main/java/hse/elysium/databaseInteractor/TokenService.java (hunk 4 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/databaseInteractor/TokenService.java:13]: It's Fedya's class

File: Elysium/src/main/java/hse/elysium/databaseInteractor/UserService.java (hunk 5 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/databaseInteractor/UserService.java:322]: Name collision

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java (hunk 6 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java]: deleted)

File: Elysium/src/main/java/hse/elysium/serverspring/config/SecurityConfig.java (hunk 12 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/config/SecurityConfig.java:43]: there are to much functions, so i decided to make ultimative solution

File: Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java (hunk 9 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java]: image

Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java]: учимся у лучших

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java (hunk 7 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/AuthController.java:19]: didn't imagine better

File: Elysium/src/main/java/hse/elysium/serverspring/auth/AuthenticationFilter.java (hunk 8 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/AuthenticationFilter.java]: now I use logger

File: Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java (hunk 10 above)
Comment (FONARMEISTER) [reply Elysium/src/main/java/hse/elysium/serverspring/auth/JwtService.java:59]: because jwt token requires extra claims

**Pull request merged**